// react-data-mapping/dist/index.js
import $ from 'jquery';
import React, { createElement, Component } from 'react';
import ReactDOM, { render, findDOMNode } from 'react-dom';
import { Canvas, Endpoint, Tips, Node } from 'butterfly-dag';
import 'butterfly-dag/dist/index.css';
import { get, isEmpty, some, assign, cloneDeep } from 'lodash';

/*! *****************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */
/* global Reflect, Promise */

var extendStatics = function(d, b) {
    extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
    return extendStatics(d, b);
};

function __extends(d, b) {
    extendStatics(d, b);
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
}

function _classCallCheck(instance, Constructor) {
  if (!(instance instanceof Constructor)) {
    throw new TypeError("Cannot call a class as a function");
  }
}

function _defineProperties(target, props) {
  for (var i = 0; i < props.length; i++) {
    var descriptor = props[i];
    descriptor.enumerable = descriptor.enumerable || false;
    descriptor.configurable = true;
    if ("value" in descriptor) descriptor.writable = true;
    Object.defineProperty(target, descriptor.key, descriptor);
  }
}

function _createClass(Constructor, protoProps, staticProps) {
  if (protoProps) _defineProperties(Constructor.prototype, protoProps);
  if (staticProps) _defineProperties(Constructor, staticProps);
  return Constructor;
}

function _defineProperty(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    });
  } else {
    obj[key] = value;
  }

  return obj;
}

function _inherits(subClass, superClass) {
  if (typeof superClass !== "function" && superClass !== null) {
    throw new TypeError("Super expression must either be null or a function");
  }

  subClass.prototype = Object.create(superClass && superClass.prototype, {
    constructor: {
      value: subClass,
      writable: true,
      configurable: true
    }
  });
  if (superClass) _setPrototypeOf(subClass, superClass);
}

function _getPrototypeOf(o) {
  _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) {
    return o.__proto__ || Object.getPrototypeOf(o);
  };
  return _getPrototypeOf(o);
}

function _setPrototypeOf(o, p) {
  _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) {
    o.__proto__ = p;
    return o;
  };

  return _setPrototypeOf(o, p);
}

function _isNativeReflectConstruct() {
  if (typeof Reflect === "undefined" || !Reflect.construct) return false;
  if (Reflect.construct.sham) return false;
  if (typeof Proxy === "function") return true;

  try {
    Date.prototype.toString.call(Reflect.construct(Date, [], function () {}));
    return true;
  } catch (e) {
    return false;
  }
}

function _assertThisInitialized(self) {
  if (self === void 0) {
    throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
  }

  return self;
}

function _possibleConstructorReturn(self, call) {
  if (call && (typeof call === "object" || typeof call === "function")) {
    return call;
  }

  return _assertThisInitialized(self);
}

function _createSuper(Derived) {
  var hasNativeReflectConstruct = _isNativeReflectConstruct();

  return function _createSuperInternal() {
    var Super = _getPrototypeOf(Derived),
        result;

    if (hasNativeReflectConstruct) {
      var NewTarget = _getPrototypeOf(this).constructor;

      result = Reflect.construct(Super, arguments, NewTarget);
    } else {
      result = Super.apply(this, arguments);
    }

    return _possibleConstructorReturn(this, result);
  };
}

var MappindCanvas = /*#__PURE__*/function (_Canvas) {
  _inherits(MappindCanvas, _Canvas);

  var _super = _createSuper(MappindCanvas);

  function MappindCanvas(opts) {
    var _this;

    _classCallCheck(this, MappindCanvas);

    _this = _super.call(this, opts);

    _defineProperty(_assertThisInitialized(_this), "_getEndpoint", function (point) {
      var _node = _this.getNode(point.nodeId);

      var _point = undefined;

      if (_node && point.originId) {
        _point = _node.getEndpoint(point.originId);
        return _point;
      } else {
        return point;
      }
    });

    _this.extraPos = opts.extraPos;
    return _this;
  }

  _createClass(MappindCanvas, [{
    key: "_calcPos",
    value: function _calcPos() {
      var sourceTop = 0 + _.get(this, 'extraPos.paddingTop', 0);

      var sourceLeft = 0 + _.get(this, 'extraPos.paddingLeft', 0);

      var souceNodes = this.nodes.filter(function (item) {
        return item.options.type === 'source';
      }).map(function (item) {
        item.moveTo(sourceLeft, sourceTop);
        sourceTop += item.height + item.PADDING_VERTICAL;
        return item;
      });

      var targetTop = 0 + _.get(this, 'extraPos.paddingTop', 0);

      var targetLeft = _.get(souceNodes, '[0].width', 0) + _.get(souceNodes, '[0].PADDING_HORIZONTAL', 0) + _.get(this, 'extraPos.paddingLeft', 0);

      this.nodes.filter(function (item) {
        return item.options.type === 'target';
      }).forEach(function (item) {
        item.moveTo(targetLeft, targetTop);
        targetTop += item.height + item.PADDING_VERTICAL;
        return item;
      });
    }
  }, {
    key: "_autoResize",
    value: function _autoResize(type) {
      var totalHeight = 0;
      var totleWidth = 0;
      var _sourceHeight = 0;
      var _targetHeight = 0;
      var souceNodes = this.nodes.filter(function (item) {
        return item.options.type === 'source';
      });
      var targetNodes = this.nodes.filter(function (item) {
        return item.options.type === 'target';
      });
      souceNodes.forEach(function (item) {
        _sourceHeight += item.height + item.PADDING_VERTICAL;
      });
      targetNodes.forEach(function (item) {
        _targetHeight += item.height + item.PADDING_VERTICAL;
      }); // 计算所有节点大小总和

      totalHeight = _sourceHeight > _targetHeight ? _sourceHeight : _targetHeight;
      totleWidth = _.get(souceNodes, '[0].width', 0) + _.get(souceNodes, '[0].PADDING_HORIZONTAL', 0) + _.get(targetNodes, '[0].width', 0) || 200; // 计算边缘

      totalHeight += _.get(this, 'extraPos.paddingTop', 0) + _.get(this, 'extraPos.paddingBottom', 0);
      totleWidth += _.get(this, 'extraPos.paddingLeft', 0) + _.get(this, 'extraPos.paddingRight', 0);

      if (type === 'width') {
        $(this.root).css('width', totleWidth);
      }

      if (type === 'height') {
        $(this.root).css('height', totalHeight);
      }

      this.updateRootResize();
    } // 纠正获取左右锚点

  }, {
    key: "_linkedChain",
    // 改变linked状态
    value: function _linkedChain(links) {
      var _this2 = this;

      links.forEach(function (edge) {
        var _sourceEndpoint = _this2._getEndpoint(edge.sourceEndpoint);

        var _targetEndpoint = _this2._getEndpoint(edge.targetEndpoint);

        _sourceEndpoint && $(_sourceEndpoint.dom).addClass('link');
        _targetEndpoint && $(_targetEndpoint.dom).addClass('link');
      });
    }
  }, {
    key: "_unLinkedChain",
    value: function _unLinkedChain(links) {
      var _this3 = this;

      links.forEach(function (edge) {
        var _sourceEndpoint = _this3._getEndpoint(edge.sourceEndpoint);

        var _targetEndpoint = _this3._getEndpoint(edge.targetEndpoint);

        _sourceEndpoint && $(_sourceEndpoint.dom).removeClass('link');
        _targetEndpoint && $(_targetEndpoint.dom).removeClass('link');
      });
    } // 聚焦链路

  }, {
    key: "_focusChain",
    value: function _focusChain(point) {
      var _this4 = this;

      var edges = this._findChain(point);

      edges.forEach(function (item) {
        _this4._changeFoucsStatus(item, true);
      });
    }
  }, {
    key: "_unFocusChain",
    value: function _unFocusChain(point) {
      var _this5 = this;

      var edges = this._findChain(point);

      edges.forEach(function (item) {
        _this5._changeFoucsStatus(item, false);
      });
    }
  }, {
    key: "_changeFoucsStatus",
    value: function _changeFoucsStatus(edge, status) {
      var _sourceEndpoint = this._getEndpoint(edge.sourceEndpoint);

      var _targetEndpoint = this._getEndpoint(edge.targetEndpoint);

      if (status) {
        $(edge.dom).addClass('focus');
        $(edge.arrowDom).addClass('focus');
        _sourceEndpoint && $(_sourceEndpoint.dom).addClass('focus');
        _targetEndpoint && $(_targetEndpoint.dom).addClass('focus');
      } else {
        $(edge.dom).removeClass('focus');
        $(edge.arrowDom).removeClass('focus');
        _sourceEndpoint && $(_sourceEndpoint.dom).removeClass('focus');
        _targetEndpoint && $(_targetEndpoint.dom).removeClass('focus');
      }
    }
  }, {
    key: "_findChain",
    value: function _findChain(point) {
      var type = point.type;
      var neighborEdges = this.getNeighborEdges(point.nodeId);
      var targetsEdges = neighborEdges.filter(function (item) {
        return item[type + 'Node'].id === point.nodeId && item[type + 'Endpoint'].originId === point.id;
      });
      return targetsEdges;
    } // 检查连接数量限制

  }, {
    key: "_checkLinkNum",
    value: function _checkLinkNum(point, targetEdge, type) {
      var _linkNums = this.edges.filter(function (_edge) {
        return _edge["".concat(type, "Node")].id === point.nodeId && _edge["".concat(type, "Endpoint")].id === point.id;
      }).length + 1;

      var _isValidLink = true;

      var _pointLimitedNum = -1;

      if (point.limitNum && typeof point.limitNum === "number") {
        if (_linkNums > point.limitNum) {
          _pointLimitedNum = point.limitNum;
          _isValidLink = false;
        }
      }

      if (point.limitNum && Object.prototype.toString.call(point.limitNum) === "[object Object]") {
        if (point.limitNum.source && type === "source") {
          if (_linkNums > point.limitNum.source) {
            _pointLimitedNum = point.limitNum.source;
            _isValidLink = false;
          }
        }

        if (point.limitNum.target && type === "target") {
          if (_linkNums > point.limitNum.target) {
            _pointLimitedNum = point.limitNum.target;
            _isValidLink = false;
          }
        }
      }

      if (!_isValidLink) {
        console.warn("id\u4E3A".concat(point.id, "\u7684\u951A\u70B9\u9650\u5236\u4E86").concat(_pointLimitedNum, "\u6761\u8FDE\u7EBF"));
        targetEdge && targetEdge.destroy();
        this._dragEdges = [];
        this._dragType = null;
      }

      return _isValidLink;
    }
  }]);

  return MappindCanvas;
}(Canvas);

var isReactEle = function isReactEle(HTMLElement) {
  return /*#__PURE__*/React.isValidElement(HTMLElement);
};
/**
 * params {Object} config
 * params {JSX.Element | String} config.content
 * params {Number | String} config.width
 */


var emptyDom = (function (config) {
  var content = config.content;
  var width = config.width;

  if (!width) {
    width = '150px';
  }

  if (typeof config.width === 'number') {
    width = config.width + 'px';
  }

  var emptyDom = '<div style="width: ' + width + '"></div>';

  if (content) {
    if (isReactEle(content)) {
      emptyDom = ReactDOM.render(content, document.createElement('div'));
    } else {
      emptyDom = $(content);
    }
  } else {
    emptyDom = $('<div class="no-data" style="width: ' + width + '"></div>');
    var iconDom = $('<i class="no-data-icon data-mapping-icon data-mapping-icon-kongshuju"></i>');
    emptyDom.append(iconDom);
  }

  return emptyDom;
});

var NewEndPoint = /*#__PURE__*/function (_Endpoint) {
  _inherits(NewEndPoint, _Endpoint);

  var _super = _createSuper(NewEndPoint);

  function NewEndPoint(opts) {
    var _this;

    _classCallCheck(this, NewEndPoint);

    _this = _super.call(this, opts);

    if (!_this.options._isNodeSelf) {
      _this.originId = (_this.id || '').replace('-left', '').replace('-right', '');
    }

    return _this;
  }

  _createClass(NewEndPoint, [{
    key: "attachEvent",
    value: function attachEvent() {
      var _this2 = this;

      $(this.dom).on('mousedown', function (e) {
        var LEFT_KEY = 0;

        if (e.button !== LEFT_KEY) {
          return;
        }

        e.preventDefault();
        e.stopPropagation(); // 点击中了上移/下移的按钮,需要阻止

        var classname = e.target.className || '';

        if (classname.indexOf('move-up') !== -1 || classname.indexOf('move-down') !== -1) {
          return;
        }

        if (_this2.options._isNodeSelf && _this2.type === 'target') {
          _this2.emit('custom.endpoint.dragNode', {
            data: _this2
          });
        } else {
          _this2.emit('InnerEvents', {
            type: 'endpoint:drag',
            data: _this2
          });
        }
      });

      if (this.options._isNodeSelf) {
        $(this.dom).on('mouseover', function (e) {
          _this2.emit('custom.endpoint.focus', {
            point: _this2
          });
        });
        $(this.dom).on('mouseout', function (e) {
          _this2.emit('custom.endpoint.unFocus', {
            point: _this2
          });
        });
      }
    }
  }]);

  return NewEndPoint;
}(Endpoint);

var TableNode = /*#__PURE__*/function (_Node) {
  _inherits(TableNode, _Node);

  var _super = _createSuper(TableNode);

  function TableNode(opts) {
    var _this;

    _classCallCheck(this, TableNode);

    _this = _super.call(this, opts); // 标题高度

    _this.TITLE_HEIGHT = 34; // 列标题高度

    _this.COLUMNS_TITLE_HEIGHT = 28; // 每列宽度

    _this.COLUMN_WIDTH = 60; // 每行高度

    _this.ROW_HEIGHT = 26; // 垂直间距

    _this.PADDING_VERTICAL = 10; // 水平间距

    _this.PADDING_HORIZONTAL = get(opts, '_extraPos.paddingCenter') || 150; // 排序宽度

    _this.SORTABLE_WIDTH = 40;
    _this.height = 0;
    _this.width = 0;
    _this.fieldsList = [];
    return _this;
  }

  _createClass(TableNode, [{
    key: "_addEventListener",
    value: function _addEventListener() {
      var _this2 = this;

      $(this.dom).on('mouseDown', function (e) {
        var LEFT_KEY = 0;

        if (e.button !== LEFT_KEY) {
          return;
        }

        if (_this2.draggable) {
          _this2._isMoving = true;

          _this2.emit('InnerEvents', {
            type: 'node:dragBegin',
            data: _this2
          });
        } else {
          // 单纯为了抛错事件给canvas，为了让canvas的dragtype不为空，不会触发canvas:click事件
          _this2.emit('InnerEvents', {
            type: 'node:mouseDown',
            data: _this2
          });

          return true;
        }
      });
    }
  }, {
    key: "mounted",
    value: function mounted() {
      this._createNodeEndpoint(); // 保持title宽度


      if (!this.fieldsList.length) {
        $(this.dom).find('.title').css('width', this.options._emptyWidth || 150);
      }

      var fieldItems = $(this.dom).find('.field-item');
      var fieldItemDoms = Array.prototype.slice.apply(fieldItems);
      fieldItemDoms.forEach(function (_fieldItem, index) {
        if (_fieldItem.scrollWidth > _fieldItem.clientWidth) {
          var fieldItem = $(fieldItems[index]);
          Tips.createTip({
            className: 'field-item-tooltip',
            targetDom: fieldItem[0],
            genTipDom: function genTipDom() {
              return fieldItem.text();
            }
          });
        }
      });
    }
  }, {
    key: "draw",
    value: function draw(obj) {
      var _dom = obj.dom;

      if (!_dom) {
        _dom = $('<div></div>').attr('class', 'node table-node').attr('id', obj.name);
      }

      if (!isEmpty(obj.options._sourceClassName) && get(obj, 'options.type') === 'source') {
        _dom.addClass(obj.options._sourceClassName);
      }

      if (!isEmpty(obj.options._targetClassName) && get(obj, 'options.type') === 'target') {
        _dom.addClass(obj.options._targetClassName);
      }

      var node = $(_dom); // 计算节点坐标

      if (obj.top !== undefined) {
        node.css('top', "".concat(obj.top + get(obj, 'options._extraPos.paddingTop', 0), "px"));
      }

      if (obj.left !== undefined) {
        node.css('left', "".concat(obj.left + get(obj, 'options._extraPos.paddingLeft', 0), "px"));
      }

      this._calcSize(node, obj);

      this._createTableName(node); // 表名


      this._createFieldTitle(node); // 字段标题


      this._createFields(node); // 字段


      return node[0];
    }
  }, {
    key: "_createTableName",
    value: function _createTableName() {
      var container = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this.dom;

      var title = get(this, 'options.title');

      if (title) {
        var titleDom = $("<div class=\"title\">".concat(title, "</div>"));
        titleDom.css({
          'height': this.TITLE_HEIGHT + 'px',
          'line-height': this.TITLE_HEIGHT + 'px'
        });
        $(container).append(titleDom);
      }
    }
  }, {
    key: "_createFieldTitle",
    value: function _createFieldTitle() {
      var _this3 = this;

      var container = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this.dom;

      var columns = get(this, 'options._columns', []);

      var hasFieldTitle = some(columns, function (item) {
        return item.title;
      });

      if (hasFieldTitle) {
        var columnsTitleDom = $('<div class="filed-title"></div>');
        columns.forEach(function (_col) {
          var columnsTitleItem = $("<span class=\"filed-title-item\">".concat(_col.title, "</span>"));
          columnsTitleItem.css('width', (_col.width || _this3.COLUMN_WIDTH) + 'px');
          columnsTitleDom.append(columnsTitleItem);
        });
        columnsTitleDom.css('height', this.COLUMNS_TITLE_HEIGHT + 'px').css('line-height', this.COLUMNS_TITLE_HEIGHT + 'px');
        container.append(columnsTitleDom);
      }
    }
  }, {
    key: "_createSortableBtn",
    value: function _createSortableBtn(field) {
      var sortFieldDom = $("\n      <span class=\"field-sort\">\n        <i class=\"data-mapping-icon data-mapping-icon-paixu-top move-up\"></i>\n        <i class=\"data-mapping-icon data-mapping-icon-paixu-bottom move-down\"></i>\n      </span>\n    ");
      sortFieldDom.css({
        width: this.SORTABLE_WIDTH + 'px'
      });
      sortFieldDom.find('.move-up').click(this._moveUp.bind(this, field));
      sortFieldDom.find('.move-down').click(this._moveDown.bind(this, field));
      return sortFieldDom;
    }
  }, {
    key: "_createFields",
    value: function _createFields() {
      var _this4 = this;

      var container = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this.dom;

      var fields = get(this, 'options.fields');

      var columns = get(this, 'options._columns', []);

      var sortable = get(this, 'options.sortable');

      var isObject = Object.prototype.toString.call(sortable) === "[object Object]";

      var type = get(this, 'options.type', '');

      if (fields && fields.length) {
        fields.forEach(function (_field, index) {
          var fieldDom = $('<div class="field"></div>');
          var _primaryKey = columns[0].key;
          var sortFieldDom = undefined;

          if (sortable && typeof sortable === 'boolean') {
            sortFieldDom = _this4._createSortableBtn(_field);
          }

          fieldDom.css({
            height: _this4.ROW_HEIGHT + 'px',
            'line-height': _this4.ROW_HEIGHT + 'px'
          });
          columns.forEach(function (_col) {
            if (_col.render) {
              var fieldItemDom = $("<span class=\"field-item\"></span>");
              fieldItemDom.css('width', (_col.width || _this4.COLUMN_WIDTH) + 'px');
              render(_col.render(_field[_col.key], _field, index), fieldItemDom[0]);
              fieldDom.append(fieldItemDom);
            } else {
              var _fieldItemDom = $("<span class=\"field-item\">".concat(_field[_col.key], "</span>"));

              _fieldItemDom.css('width', (_col.width || _this4.COLUMN_WIDTH) + 'px');

              fieldDom.append(_fieldItemDom);
            }

            if (_col.primaryKey) {
              _primaryKey = _col.key;
            }
          });

          if (sortFieldDom) {
            fieldDom.append(sortFieldDom);
          }

          if (type === 'source') {
            var rightPoint = $('<div class="point right-point"></div>');
            fieldDom.append(rightPoint);

            if (isObject && sortable.source) {
              sortFieldDom = _this4._createSortableBtn(_field);
              fieldDom.append(sortFieldDom);
            }
          }

          if (type === 'target') {
            var leftPoint = $('<div class="point left-point"></div>');
            fieldDom.append(leftPoint);

            if (isObject && sortable.target) {
              sortFieldDom = _this4._createSortableBtn(_field);
              fieldDom.append(sortFieldDom);
            }
          }

          container.append(fieldDom);

          _this4.fieldsList.push({
            id: _field[_primaryKey],
            dom: fieldDom
          });
        });
      } else {
        var _emptyContent = get(this.options, '_emptyContent');

        var noDataTree = emptyDom({
          content: _emptyContent,
          width: this.options._emptyWidth
        });
        container.append(noDataTree);
      }
    }
  }, {
    key: "_createNodeEndpoint",
    value: function _createNodeEndpoint() {
      var _this5 = this;

      var type = this.options.type;
      this.fieldsList.forEach(function (item) {
        _this5.addEndpoint({
          id: item.id,
          orientation: type === 'source' ? [1, 0] : [-1, 0],
          type: type,
          _isNodeSelf: true,
          dom: item.dom[0],
          Class: NewEndPoint
        });

        if (type === 'source') {
          _this5.addEndpoint({
            id: item.id + '-right',
            orientation: [1, 0],
            type: type,
            _isNodeSelf: false,
            dom: $(item.dom).find('.right-point')[0],
            Class: NewEndPoint,
            linkable: false
          });
        } else if (type === 'target') {
          _this5.addEndpoint({
            id: item.id + '-left',
            orientation: [-1, 0],
            type: type,
            _isNodeSelf: false,
            dom: $(item.dom).find('.left-point')[0],
            Class: NewEndPoint,
            disLinkable: false
          });
        }
      });
    }
  }, {
    key: "_calcSize",
    value: function _calcSize(node, obj) {
      var _this6 = this;

      var hasTitle = get(obj, 'options.title');

      var fields = get(obj, 'options.fields', []);

      var sortable = get(obj, 'options.sortable');

      var type = get(obj, 'options.type');

      if (hasTitle) {
        this.height += this.TITLE_HEIGHT;
      }

      this.height += fields.length * this.ROW_HEIGHT;

      var columns = get(obj, 'options._columns');

      columns.forEach(function (item) {
        _this6.width += item.width || _this6.COLUMN_WIDTH;
      });
      if (typeof sortable === 'boolean') this.width += this.SORTABLE_WIDTH;

      if (Object.prototype.toString.call(sortable) === '[object Object]') {
        if (type === 'source') this.width += this.SORTABLE_WIDTH;
        if (type === 'target') this.width += this.SORTABLE_WIDTH;
      } // todo: 记得算上SORTABLE_WIDTH

    }
  }, {
    key: "_moveUp",
    value: function _moveUp(curField, event) {
      event.preventDefault();
      event.stopPropagation();
      var curIndex = this.fieldsList.findIndex(function (i) {
        return i.id === curField.id;
      });

      var oldFields = get(this, 'options.fields', []);

      var oldFieldsItem = oldFields.splice(curIndex, 1);
      var point = this.getEndpoint(curField.id);
      var curFieldDom = point.dom;
      var curFieldData = this.fieldsList[curIndex]; // 处理边界

      if (curIndex === 0) {
        console.warn('this field has reach the top!');
        return;
      }

      var preFieldData = this.fieldsList[curIndex - 1];
      var preFieldDom = preFieldData.dom; // 交换dom

      $(preFieldDom).before(curFieldDom); // 交换数据

      this.fieldsList[curIndex] = preFieldData;
      this.fieldsList[curIndex - 1] = curFieldData;
      oldFields.splice(curIndex - 1, 0, oldFieldsItem[0]); // 发送事件，更新线段和锚点坐标

      this.emit('custom.field.sort', {
        nodeId: this.id,
        pointIds: [curFieldData.id, preFieldData.id]
      });
    }
  }, {
    key: "_moveDown",
    value: function _moveDown(curField, event) {
      event.preventDefault();
      event.stopPropagation();
      var curIndex = this.fieldsList.findIndex(function (i) {
        return i.id === curField.id;
      });

      var oldFields = get(this, 'options.fields', []);

      var oldFieldsItem = oldFields.splice(curIndex, 1);
      var point = this.getEndpoint(curField.id);
      var curFieldDom = point.dom;
      var curFieldData = this.fieldsList[curIndex]; // 处理边界

      if (curIndex === this.fieldsList.length - 1) {
        console.warn('this field has reach the bottom!');
        return;
      }

      var nextFieldData = this.fieldsList[curIndex + 1];
      var nextFieldDom = nextFieldData.dom; // 交换dom

      $(nextFieldDom).after(curFieldDom); // 交换数据

      this.fieldsList[curIndex] = nextFieldData;
      this.fieldsList[curIndex + 1] = curFieldData;
      oldFields.splice(curIndex + 1, 0, oldFieldsItem[0]); // 发送事件，更新线段和锚点坐标

      this.emit('custom.field.sort', {
        nodeId: this.id,
        pointIds: [curFieldData.id, nextFieldData.id]
      });
    }
  }]);

  return TableNode;
}(Node);

var transformInitData = function transformInitData(data) {
  var columns = data.columns,
      sourceData = data.sourceData,
      targetData = data.targetData,
      mappingData = data.mappingData,
      type = data.type,
      extraPos = data.extraPos,
      sortable = data.sortable,
      emptyContent = data.emptyContent,
      emptyWidth = data.emptyWidth,
      sourceClassName = data.sourceClassName,
      targetClassName = data.targetClassName;

  var _genNodes = function _genNodes(data, nodeType, comType) {
    if (comType === 'single' && data.constructor === Object) {
      return [assign({
        id: nodeType,
        type: nodeType,
        _columns: columns,
        _extraPos: extraPos,
        Class: TableNode,
        _emptyContent: emptyContent,
        _emptyWidth: emptyWidth,
        _sourceClassName: sourceClassName,
        _targetClassName: targetClassName,
        sortable: sortable
      }, data)];
    } else if (comType === 'mutiply' && data.constructor === Array) {
      return data.map(function (item) {
        return assign({
          type: nodeType,
          _columns: columns,
          _extraPos: extraPos,
          Class: TableNode,
          _emptyContent: emptyContent,
          _emptyWidth: emptyWidth,
          _sourceClassName: sourceClassName,
          _targetClassName: targetClassName,
          sortable: sortable
        }, item);
      });
    }
  };

  var sourceNodes = _genNodes(sourceData, 'source', type);

  var targetNodes = _genNodes(targetData, 'target', type);

  var edges = mappingData.map(function (item) {
    return {
      id: "".concat(item.source, "-").concat(item.target),
      type: 'endpoint',
      sourceNode: item.sourceNode || sourceNodes[0].id,
      source: item.source,
      targetNode: item.targetNode || targetNodes[0].id,
      target: item.target,
      _isConnect: data.isConnect
    };
  });
  return {
    nodes: [].concat(sourceNodes).concat(targetNodes),
    edges: edges
  };
};
var transformChangeData = function transformChangeData(data, comType) {
  var result = {
    mappingData: [],
    sourceData: [],
    targetData: []
  };
  var sourceNodes = data.nodes.filter(function (item) {
    return item.options.type === 'source';
  });
  var targetNodes = data.nodes.filter(function (item) {
    return item.options.type === 'target';
  });

  if (comType === 'single') {
    var _sourceNode = sourceNodes[0];
    var _targetNode = targetNodes[0];
    result.mappingData = data.edges.map(function (item) {
      return {
        source: item.sourceEndpoint.originId,
        target: item.targetEndpoint.originId
      };
    });
    result.sourceData = {
      title: get(_sourceNode, 'options.title'),
      fields: get(_sourceNode, 'options.fields')
    };
    result.targetData = {
      title: get(_targetNode, 'options.title'),
      fields: get(_targetNode, 'options.fields')
    };
  } else if (comType === 'mutiply') {
    result.mappingData = data.edges.map(function (item) {
      return {
        sourceNode: item.sourceNode.id,
        targetNode: item.targetNode.id,
        source: item.sourceEndpoint.originId,
        target: item.targetEndpoint.originId
      };
    });
    result.sourceData = sourceNodes.map(function (item) {
      return {
        title: get(item, 'options.title'),
        fields: get(item, 'options.fields')
      };
    });
    result.targetData = targetNodes.map(function (item) {
      return {
        title: get(item, 'options.title'),
        fields: get(item, 'options.fields')
      };
    });
  }

  return result;
};

var DataMapping = /** @class */ (function (_super) {
    __extends(DataMapping, _super);
    function DataMapping(props) {
        var _this = _super.call(this, props) || this;
        _this.canvas = null;
        return _this;
    }
    DataMapping.prototype.componentDidMount = function () {
        var _this = this;
        var root = findDOMNode(this);
        if (this.props.width !== undefined || this.props.width !== 'auto') {
            root.style.width = (this.props.width || 500) + 'px';
        }
        if (this.props.height !== undefined || this.props.height !== 'auto') {
            root.style.height = (this.props.height || 500) + 'px';
        }
        var result = transformInitData({
            columns: this.props.columns,
            type: this.props.type || 'single',
            sortable: get(this.props, 'config.sortable') || false,
            sourceData: cloneDeep(this.props.sourceData),
            targetData: cloneDeep(this.props.targetData),
            mappingData: cloneDeep(this.props.mappingData),
            extraPos: get(this.props, 'config.extraPos'),
            linkNumLimit: get(this.props, 'config.linkNumLimit'),
            emptyContent: this.props.emptyContent,
            emptyWidth: this.props.emptyWidth,
            sourceClassName: this.props.sourceClassName || '',
            targetClassName: this.props.targetClassName || ''
        });
        var canvasObj = {
            root: root,
            disLinkable: true,
            linkable: true,
            draggable: false,
            zoomable: false,
            moveable: false,
            theme: {
                edge: {
                    type: 'AdvancedBezier',
                    arrow: true,
                    isExpandWidth: true,
                    arrowPosition: 1,
                    arrowOffset: 5
                },
                endpoint: {
                    limitNum: undefined,
                    expandArea: {
                        left: 0,
                        right: 0,
                        top: 0,
                        botton: 0
                    }
                }
            },
            extraPos: get(this.props, 'config.extraPos')
        };
        var _linkNumLimit = get(this.props, 'config.linkNumLimit');
        if (typeof _linkNumLimit === 'number' && !isNaN(_linkNumLimit)) {
            canvasObj.theme.endpoint.limitNum = _linkNumLimit;
        }
        if (Object.prototype.toString.call(_linkNumLimit) === '[object Object]') {
            canvasObj.theme.endpoint.limitNum = _linkNumLimit;
        }
        this.canvas = new MappindCanvas(canvasObj);
        setTimeout(function () {
            _this.canvas.draw(result, function () {
                _this.canvas._calcPos();
                if (_this.props.width === 'auto') {
                    _this.canvas._autoResize('width');
                }
                if (_this.props.height === 'auto') {
                    _this.canvas._autoResize('height');
                }
                _this.props.onLoaded && _this.props.onLoaded(_this.canvas);
            });
            _this._addEventListener();
        }, get(this.props, 'config.delayDraw', 0));
    };
    DataMapping.prototype.onChange = function () {
        var result = transformChangeData(this.canvas.getDataMap(), this.props.type || 'single');
        this.props.onChange && this.props.onChange(result);
    };
    DataMapping.prototype._genClassName = function () {
        var classname = '';
        if (this.props.className) {
            classname = this.props.className + ' butterfly-data-mapping';
        }
        else {
            classname = 'butterfly-data-mapping';
        }
        return classname;
    };
    DataMapping.prototype._addEventListener = function () {
        var _this = this;
        var _addLinks = function (links) {
            var newLinkOpts = links.map(function (item) {
                var _oldSource = get(item, 'sourceEndpoint.id', '');
                var _oldTarget = get(item, 'targetEndpoint.id', '');
                var _newSource = _oldSource.indexOf('-right') !== -1 ? _oldSource : _oldSource + '-right';
                var _newTarget = _oldTarget.indexOf('-left') !== -1 ? _oldTarget : _oldTarget + '-left';
                return {
                    id: item.options.id,
                    sourceNode: item.options.sourceNode,
                    targetNode: item.options.targetNode,
                    source: _newSource,
                    target: _newTarget,
                    type: 'endpoint'
                };
            });
            _this.canvas.removeEdges(links, true);
            newLinkOpts = newLinkOpts.filter(function (item) {
                var targetNode = _this.canvas.getNode(item.targetNode);
                var targetEndpoint = targetNode.getEndpoint(item.target);
                var sourceEndpoint = targetNode.getEndpoint(item.source);
                var result = _this.canvas._checkLinkNum(targetEndpoint, undefined, 'target');
                // 取消link状态
                if (!result) {
                    sourceEndpoint && $(sourceEndpoint.dom).removeClass('link');
                }
                return result;
            });
            return _this.canvas.addEdges(newLinkOpts, true);
        };
        var _isInit = true;
        this.canvas.on('system.link.connect', function (data) {
            var addEdges = _addLinks(data.links || []);
            var result = [];
            addEdges.forEach(function (item) {
                var isConnect = true;
                _this.props.isConnect && (isConnect = _this.props.isConnect(item));
                if (isConnect) {
                    result.push(item);
                }
                else {
                    _this.canvas.removeEdge(item, true);
                }
            });
            if (!_isInit) {
                _this.onChange();
            }
            _isInit = false;
            _this.canvas._linkedChain(result);
        });
        this.canvas.on('system.link.reconnect', function (data) {
            var addEdges = _addLinks(data.addLinks || []);
            var result = [];
            addEdges.forEach(function (item) {
                var isConnect = true;
                _this.props.isConnect && (isConnect = _this.props.isConnect(item));
                if (isConnect) {
                    result.push(item);
                }
                else {
                    _this.canvas.removeEdge(item, true);
                }
            });
            _this.onChange();
            _this.canvas._unLinkedChain(data.delLinks);
            _this.canvas._linkedChain(result);
        });
        this.canvas.on('system.links.delete', function (data) {
            _this.onChange();
            _this.canvas._unLinkedChain(data.links);
        });
        // 线段删除特殊处理
        this.canvas.on('custom.endpoint.dragNode', function (data) {
            var point = data.data;
            var node = _this.canvas.getNode(point.nodeId);
            var linkedPoint = node.getEndpoint(point.id + '-left', 'target');
            _this.canvas.emit('InnerEvents', {
                type: 'endpoint:drag',
                data: linkedPoint
            });
        });
        // 连线特殊处理
        this.canvas.on('system.drag.move', function (data) {
            var dragEdge = get(data, 'dragEdges[0]');
            var sourcePointId = get(dragEdge, 'sourceEndpoint.id', '');
            if (sourcePointId.indexOf('right') === -1) {
                var souceNode = get(dragEdge, 'sourceNode');
                var newSourcePoint = souceNode.getEndpoint(sourcePointId + '-right');
                dragEdge.sourceEndpoint = newSourcePoint;
                dragEdge.options.sourceEndpoint = newSourcePoint;
                _this.canvas._checkLinkNum(newSourcePoint, dragEdge, 'source');
            }
        });
        // 聚焦链路
        this.canvas.on('custom.endpoint.focus', function (data) {
            _this.canvas._focusChain(data.point);
            _this.props.onRowMouseOver && _this.props.onRowMouseOver(data.point);
        });
        // 失焦链路
        this.canvas.on('custom.endpoint.unFocus', function (data) {
            _this.canvas._unFocusChain(data.point);
            _this.props.onRowMouseOut && _this.props.onRowMouseOut(data.point);
        });
        // 字段重新排列
        this.canvas.on('custom.field.sort', function (data) {
            var nodeId = data.nodeId, pointIds = data.pointIds;
            var node = _this.canvas.getNode(nodeId);
            if (!node) {
                return;
            }
            pointIds.forEach(function (pointId) {
                var fieldPoints = [
                    node.getEndpoint(pointId),
                    node.getEndpoint(pointId + '-left'),
                    node.getEndpoint(pointId + '-right')
                ];
                fieldPoints.forEach(function (point) {
                    if (!point) {
                        return;
                    }
                    point.updatePos();
                });
                var updateEdges = _this.canvas.edges.filter(function (item) {
                    if (nodeId === item.sourceNode.id && (pointId + '-right' === item.sourceEndpoint.id)) {
                        return true;
                    }
                    if (nodeId === item.targetNode.id && (pointId + '-left' === item.targetEndpoint.id)) {
                        return true;
                    }
                    return false;
                });
                updateEdges.forEach(function (item) {
                    item.redraw();
                });
                _this.onChange();
            });
        });
    };
    DataMapping.prototype.render = function () {
        return (createElement("div", { className: this._genClassName() }));
    };
    return DataMapping;
}(Component));

export default DataMapping;
